README.md
              ***** Usage *****
              =================

The program can be used as a special musical instrument
or a voicing for Ukrainian text. It is not
phonetically exact and proper, but it gives
an inspiration to listen to the Ukrainian sounds more attentively.
It can be used by writers (especially poets) to compose melodic
texts. It can be used simply for the experiments. There is an
opinion that Ukrainian is a very melodic language, so you can
even not knowing it get some impression and get acquainted with
this beautiful language. There are a lot of songs in it and the
program is intended to open some part of beauty of the language itself.

The phonetic and writing system material are from the source:

[Solomija Buk, Ján Mačutek, Andrij Rovenchak. Some properties of
the Ukrainian writing system. Electronic resource](https://arxiv.org/ftp/arxiv/papers/0802/0802.4198.pdf)

Since the version 0.9.0.0 switched to the ukrainian-phonetics-basic package.

Please, update all the dependent packages related to the mmsyn6ukr to use the new approach.


ChangeLog.md
# Revision history for mmsyn6ukr

## 0.1.0.0 -- 2019-11-25

* First version. Released on an unsuspecting world.

## 0.2.0.0 -- 2019-11-25

* Second version. Added the possibility for program to use the installed
SoX binary in the system to automatically create a raw PCM file into
the .wav sound file that can be easily listened to.

## 0.2.0.1 -- 2019-11-25

* Second version revised A. Some minor code changes.

## 0.3.0.0 -- 2019-11-28

* Third version. Fixed an issue with the unnecessary additional file.
Changed the package into executable and library format.

## 0.3.0.1 -- 2019-11-28

* Third version revised A. Fixed an issue with the bounds for the library in the .cabal file.

## 0.3.1.0 -- 2019-11-29

* Third version revised B. Changed some sounds or sound representation: "и", "й", "в", "ь", "ць".

## 0.3.2.0 -- 2019-11-29

* Third version revised C. Changed some sounds or sound representation: "с", "сь", "х", "ч", "ш".

## 0.3.3.0 -- 2019-12-03

* Third version revised D. Changed all non-silent sounds. Added support for '-'.

## 0.3.4.0 -- 2019-12-03

* Third version revised E. Fixed issues with the jotted sounds.

## 0.4.0.0 -- 2019-12-05

* Fourth version. Fixed issues with the sound conversions. Now they works properly.

## 0.4.0.1 -- 2019-12-06

* Fourth version revised A. Better documented. Some minor code changes.

## 0.4.0.2 -- 2019-12-06

* Fourth version revised B. Fixed issues with the documentation.

## 0.4.0.3 -- 2019-12-06

* Fourth version revised C. Fixed issues with the documentation.

## 0.4.0.4 -- 2019-12-06

* Fourth version revised D. Fixed issues with the documentation. Fixed an issue with the nSymbols function in the Melodics.Ukrainian module.

## 0.4.0.5 -- 2019-12-06

* Fourth version revised E. Improved the documentation.

## 0.4.0.6 -- 2019-12-06

* Fourth version revised F. Fixed an issue with the documentation for Main.main function.

## 0.4.1.0 -- 2019-12-06

* Fourth version revised G. Improved the documentation for the Main.main function. Added the possibility to automatically delete the unneeded .raw file.

## 0.4.2.0 -- 2019-12-06

* Fourth version revised H. The program if there is SoX installed in the system converts the .raw file to .wav file and deletes the first one to save space.

## 0.4.2.1 -- 2019-12-06

* Fourth version revised I. Minor improvements to documentation.

## 0.4.2.2 -- 2019-12-06

* Fourth version revised J. Added an informative message after the creation of the .raw file.

## 0.5.0.0 -- 2019-12-07

* Fifth version. Changed a behaviour of the Main.main function, added a possibility of more control on the program mmsyn6ukr.

## 0.5.1.0 -- 2019-12-09

* Fifth version revised A. Fixed the genControl function, added more documentation, make the exception message more exact.

## 0.5.2.0 -- 2019-12-16

* Fifth version revised B. Added support for better pauses and therefore additional file "-.wav". Fixed issues with the conversion in case of punctuation separated words.
Added additional information about sound files. Deprecated (===) operator in favour of `elem` function.

## 0.6.0.0 -- 2019-12-18

* Sixth version. Changed the module structure.

## 0.6.1.0 -- 2019-12-24

* Sixth version revised A. Changed the dependencies bounds so that it can now be compiled for the GHC 8.8.1.

## 0.6.2.0 -- 2019-12-24

* Sixth version revised B. Changed the imported functions in the Melodics.Ukrainian module.

## 0.6.3.0 -- 2020-01-31

* Sixth version revised C. Changed a README.md file to README.markdown. Some documentation improvements.

## 0.6.3.1 -- 2020-02-23

* Sixth version revised D. Fixed an issue with word usage (a terminology) in the documentation for 'genControl' function.

## 0.6.4.0 -- 2020-05-11

* Sixth version revised E. Changed sound representations for the Ukrainian "и" -- file "K.wav", "к" -- file "M.wav", "т" -- file "V.wav", "ч" -- file "b.wav",
"і" -- file "e.wav".

## 0.6.5.0 -- 2020-05-12

* Sixth version revised F. Changed sound representations for the Ukrainian "б" -- file "B.wav", "д" -- file "E.wav", "дж" -- file "F.wav",
"ж" -- file "I.wav", "п" -- file "R.wav", "р" -- file "S.wav". Changed durations accordingly in the Map.txt file.

## 0.6.6.0 -- 2020-05-12

* Sixth version revised G. Changed sound representations for the Ukrainian "б" -- file "B.wav", "к" -- file "M.wav".

## 0.7.0.0 -- 2020-05-12

* Seventh version. Changed the module structure (and some functions accordinly, too) to more logical and convenient for factorization and composition.
Added a new module Melodics.Executable with the most of the Main module processment functions. Changed the behaviour of the executable
into repeated so you can create many files with it during one runtime operation.

## 0.7.1.0 -- 2020-05-12

* Seventh version revised A. Fixed issues with being not circled because of the semi-closed handles.

## 0.7.2.0 -- 2020-05-14

* Seventh version revised B. Fixed bounds of the dependencies so that now it is supported also for GHC 8.10* series.

## 0.7.3.0 -- 2020-06-01

* Seventh version revised C. Added Melodics.Executable.recFileName function to the export list of the module because it is re-used by the dobutokO-poetry
package.

## 0.8.0.0 -- 2020-08-16

* Eighth version. Added a new module MMSyn6Ukr.Show7s with the function from mmsyn7s module. It is done for the possible refactoring and changing the dependencies
for other packages (first of all, mmsyn7h).

## 0.8.1.0 -- 2020-09-25

* Eighth version revised A. Added to the exported functions isUkrainian from the Melodics.Ukrainian module.

## 0.8.2.0 -- 2020-10-28

* Eighth version revised B. Fixed issues with some combinations that were not
properly handled.

## 0.8.3.0 -- 2020-10-29

* Eighth version revised C. Changed the dependencies boundaries. Some minor code improvements.

## 0.9.0.0 -- 2020-11-07

* Ninth version. Switched to the ukrainian-phonetics-basic package (and so added it as a new dependency) as the engine of the phonetics conversions and

Main.hs
-- |
-- Module      :  Main
-- Copyright   :  (c) OleksandrZhabenko 2019-2020
-- License     :  MIT
-- Maintainer  :  olexandr543@yahoo.com
--
-- A program and a library that can be used as a musical instrument synthesizer or for Ukrainian speech synthesis 
-- especially for poets, translators and writers. 
--

module Main where

import System.Environment (getArgs)
import Melodics.Executable

{-
-- Inspired by: https://mail.haskell.org/pipermail/beginners/2011-October/008649.html
-}

-- | The function creates a raw PCM sound file with bitrate 22050 Hz 1 mono channel 16-bit signed-integer encoding 
-- and tries to automatically convert it to the .wav, .ogg, or .flac file with the same parameters specified by the first command line argument
-- (for more details see: 'genControl' function) using the system binary SoX (this is done for one circle of running, afterwards it is repeated 
-- with the same command line arguments. To stop execution, please, interrupt the program e. g. with Ctrl + C on many Unix platforms). 
-- So actually, it can create multiple sound files, all in the same format options specified by the first command line argument accordingly to 
-- the 'genControl' function.
--
-- If SoX binaries are not installed properly, the program makes ending informational message for the user. 
-- 
-- The command line argument is described in more details in the documentation for the 'Melodics.Ukrainian.nSymbols' function. 
-- 
-- * Notification.
-- 
-- Please, notice that successful usage of the SoX installed in the system at the moment of running the @mmsyn6ukr@ can lead to approximately doubling 
-- (in the most space consuming variant) the size of used space in the storage for the resulting files while being processed because it adds a header 
-- to the .raw file and writes down additionally the raw data to form a .wav file. Afterwards, it deletes the .raw file, so space is used finally in the 
-- more efficient manner. 
-- 
-- Also notice that the size of the largest data file representing a symbol or their combination is 6792 bytes (with 44-byte header included). So, if you 
-- expect to create sounding for @n@ symbols of the Ukrainian text, provide at least @2 * (6792 - 44) * n + 44 = 13496 * n + 44@ (bytes) of the additional 
-- space in the storage (in reality it can occupy much less because other data files are less in size). Afterwards, the program deletes the .raw file (this 
-- will approximately halve the occupied space by the resulting file) and you can manually compress the .wav file (e. g. FLAC compression with the best ratio 
-- gives approximately 0.53 of the original size. Therefore, the resulting file for the @mmsyn6ukr@ executable run prior to such operations without command 
-- line arguments is expected to be less than about @10^7@ bytes that is about 100 MB (for 31416 symbols Ukrainian text)). 
-- 
-- The best comression ratio is with the .ogg files, but they lose some quality so be careful if you need it.
main :: IO ()
main = do
  zs <- fmap (concat . take 1) getArgs
  circle zs


MMSyn6Ukr/Show7s.hs
-- |
-- Module      :  MMSyn6Ukr.Show7s
-- Copyright   :  (c) OleksandrZhabenko 2020
-- License     :  MIT
-- Stability   :  Experimental
-- Maintainer  :  olexandr543@yahoo.com
--
-- Can be used to show a sorted list of the Ukrainian sounds 
-- representations that for mmsyn7 series of programs. Is taken from 
-- the mmsyn7s package.
--

module MMSyn6Ukr.Show7s (
  show7s
) where

import qualified Data.Vector as V
import Data.List (sort, nub)
import Melodics.Ukrainian (convertToProperUkrainian)

-- | Function takes a Ukrainian text being a @String@ and returns a sorted list of the Ukrainian sounds representations that can be used further in mmsyn7 series of
-- programs.
show7s :: String -> [String]
show7s = sort . nub . V.toList . V.filter (\x -> x /= "-" && x /= "1" && x /= "0") . convertToProperUkrainian


UkrainianLControl.hs
-- |
-- Module      :  UkrainianLControl
-- Copyright   :  (c) OleksandrZhabenko 2019-2020
-- License     :  MIT
-- Maintainer  :  olexandr543@yahoo.com
--
-- A program and a library that can be used as a musical instrument synthesizer or for Ukrainian speech synthesis 
-- especially for poets, translators and writers. 
--

module UkrainianLControl (
  -- * Control the program
  genControl,
  -- * Security and Limits
  nSymbols
) where

import Data.Char (isDigit)
import qualified Data.Vector as V (generate)
import CaseBi (getBFst')

-- | Function that converts the first digit in the command line argument (starting the argument or being the second one after the letter character) given, 
-- which is a digit in the range @[0..9]@ (providing an ascending approximately exponential scale with a basis of 10 starting from 2 and ending at 1000000001), 
-- to the upper bound of number of symbols that the 'main' function of the @mmsyn6ukr@ executable reads from the 'System.IO.stdin' for sounding.
-- The default resulting value (no input) is 31416. If there is another first command line argument then the program 
-- terminates with the informational message. Using the command line argument is done for the security reasons: 
-- because of performative writing to the resulting file(s) there is a need to limit the used memory. For most cases it is
-- enough to use the default value. If you have enough resources and a large Ukrainian text amount then specify the higher values 
-- (5 or a greater one). 
nSymbols :: String -> Int
nSymbols xs | null xs = 31416::Int
            | otherwise = getBFst' (31416::Int, V.generate 10 (\n -> (n, (10^n + 1)::Int))) (let temp = read xs::Int in if temp <= 9 && temp >= 0
                            then temp 
                            else error "Please, specify a digit as a command line argument for the program!")

-- | Function that prepares arguments for the controlling functions for the executable @mmsyn6ukr@. It takes a first command line argument and makes 
-- an analysis to produce a set of String. The first resulting String is an argument to 'nSymbols' function, the first in the inner tuple is an argument
-- to the compression level for the comressed formats and the last one is the resulting file extension. The default value (no command line arguments) is
-- @("", ("", ".wav"))@. Please, specify the command line argument (if needed) in the form \"ABC\""
-- where A is either a letter \'f\', \'o\', \'w\' or a digit and B and C are both digits (or something equivalent, see below). 
-- 
-- Their meaning:
-- 
-- A:
-- 
-- \'f\' -> native FLAC format with compression from 0 (least) to 8 (best compression ratio) specified by the third characters; \'9\' is equivalent to \'8\'. This format is optional so, 
-- please, check whether it is supported by your SoX binaries. If no, install the SoX with support for the format. For more information, please, refer to the @sox@ documentation.
-- 
-- \'o\' -> Ogg Vorbis format with compression from -1 (best) to 10 (least) specified by the characters after the first two characters. The default value is "-1". This format is optional 
-- so, please, check whether it is supported by your SoX binaries. If no, install the SoX with support for the format. For more information, please, refer to the @sox@ documentation.
-- 
-- \'w\' -> WAV format with two options for rate - 11025 Hz if the third character is less than '5' and greater than '0' and otherwise 22050 Hz (the default one also for no command line arguments). 
-- 
-- If A is a digit, then it is used accordingly to 'nSymbols' function and SoX (if properly installed) quickly converts the .raw file to the default .wav with 22050 Hz rate.
-- 
-- To obtain the best compression ratio, please specify something like \"o9-1\" or \"o5-1\" (or similar). For the best lossless compression - \"f98\" or \"f58\" (or similar). 
-- 
-- For more information, please, see the @sox@ manuals (e. g. for @soxformat@).
genControl :: String -> (String, (String, String))
genControl (x:xs) | x == 'f' = ([head xs], ("-C" ++ (if (compare (tail xs) "9" /= GT) && (compare (tail xs) "0" /= LT) then take 1 . tail $ xs else "8"), ".flac"))
                  | x == 'o' = ([head xs], ("-C" ++ (if ((compare (tail xs) "9" /= GT) && (compare (tail xs) "0" /= LT)) then take 1 . tail $ xs else if (tail xs == "10") then "10" else "-1"), ".ogg"))
                  | x == 'w' = ([head xs], ("-r" ++ (if ((compare (tail xs) "4" /= GT) && (compare (tail xs) "0" /= LT)) then "11025" else "22050"), ".wav"))
                  | isDigit x = ([x], ("", ".wav"))
                  | otherwise = ("", ("", ".wav"))
genControl [] = ("", ("", ".wav"))


Melodics/Executable.hs
-- |
-- Module      :  Melodics.Executable
-- Copyright   :  (c) OleksandrZhabenko 2019-2020
-- License     :  MIT
-- Maintainer  :  olexandr543@yahoo.com
--
-- A program and a library that can be used as a musical instrument synthesizer or for Ukrainian speech synthesis
-- especially for poets, translators and writers.
--

module Melodics.Executable (
  circle
  , workWithInput
  , rawToSoundFile
  , printInfoF
  , recFileName
)
where

import Data.Char (isSpace, isControl)
import Data.Maybe (isJust,fromJust)
import System.IO
import System.Process (callProcess)
import System.Directory (removeFile)
import Control.Exception (onException)
import EndOfExe (showE)
import Melodics.Ukrainian (appendS16LEFile, convertToProperUkrainian)
import UkrainianLControl

-- | Is used to repeat the cycle of creation of the sound files in the current directory for the @mmsyn6ukr@  executable.
circle :: String -> IO ()
circle zs = onException (mapM_ (workWithInput zs) [1..]) (do
    putStr "Notice, there was (may be) CmdLineArgument exception. To avoid it, please, specify the command line argument (if needed) in the form \"ABC\""
    putStr " where A is either a letter \'f\', \'o\', \'w\' or a digit and B and C are both digits! The exception (may be) arose from "
    putStrLn $ "the command line arguments " ++ show zs ++ ". Please, check also whether the SoX was installed with the support for needed codec.")

-- | Interactively creates sound files in the current directory for the Ukrainian text input. Is used internally in the 'circle'
workWithInput :: String -> Int -> IO ()
workWithInput zs _ = do
  [nameSF,ys] <- nameAndControl zs [1,2]
  withBinaryFile (nameSF ++ ".raw") AppendMode (appendS16LEFile (convertToProperUkrainian ys))
  putStrLn "The .raw file was created by the program. If there is SoX installed then it will run further. "
  let ts = showE "sox"
  if isJust ts
    then rawToSoundFile zs nameSF (fromJust ts)
    else printInfoF

-- | Is used to retriev the user-defined file name for the record.
recFileName :: IO String
recFileName = do
  putStrLn "Please, specify the name of the resulting sound file. Please, do NOT use '}' character and space or control characters!"
  nameOfSoundFile <- getLine
  let nameSF = filter (\x -> not (isSpace x) && not (isControl x) && x /= '}') nameOfSoundFile
  return nameSF

getCtrl :: String -> IO String
getCtrl zs = do
  xs <- getLine
  let ys = take (nSymbols . fst . genControl $ zs) xs
  return ys

recFNAndCtrl :: String -> Int -> IO String
recFNAndCtrl zs n
  | odd n = recFileName
  | otherwise = getCtrl zs

nameAndControl :: String -> [Int] -> IO [String]
nameAndControl zs = mapM (recFNAndCtrl zs)

-- | Converts RAW sound to the sound file of the needed format in the current directory accordingly to the 'genControl' for the first 'String' argument.
-- Is used internally in the 'workWithInput'.
rawToSoundFile :: String -> String -> FilePath -> IO ()
rawToSoundFile zs nameSF executablePath
  | null zs = do
     callProcess executablePath ["-r22050","-c1","-L","-esigned-integer","-b16", nameSF ++ ".raw", nameSF ++ ".wav"]
     removeFile $ nameSF ++ ".raw"
  | otherwise = do
     let ws = snd . genControl $ zs
     callProcess executablePath ["-r22050","-c1","-L","-esigned-integer","-b16", nameSF ++ ".raw", fst ws, nameSF ++ snd ws]
     removeFile $ nameSF ++ ".raw"

-- | Prints informational message about ending of the possible for the given data program operation on sound files. Is used internally in the 'workWithInput'.
-- Is used internally in the 'workWithInput'.
printInfoF :: IO ()
printInfoF = do
  putStr "You have a resulting file in a raw PCM format with bitrate 22050 Hz and 1 channel (mono) in the .raw format. "
  putStr "You can further process it by yourself manually, otherwise, please, install FFMpeg or LibAV executables in the directory mentioned in the variable PATH"
  putStrLn " and then run: "
  putStrLn "\"name_of_FFMpeg_or_LibAV_executable\" -f s16le -acodec pcm_s16le -ac 1 -ar 22050 -i \"name_Of_the_sound_file\" \"the_same_name_without_.raw_ending_and_with_.wav_ending\""
  putStrLn ""
  putStrLn "OR you can install SoX executable in the directory mentioned in the variable PATH and then run: "
  putStrLn "\"Path_to_the_SoX_executable\" -b16 -r22050 -c1 -e signed-integer -L \"name_of_the_file_in_raw_format_with_new._prefix\" \"name_of_the_file_in_raw_format_with_new._prefix\" in the terminal."


Melodics/Ukrainian.hs
-- |
-- Module      :  Melodics.Ukrainian
-- Copyright   :  (c) OleksandrZhabenko 2019-2020
-- License     :  MIT
-- Maintainer  :  olexandr543@yahoo.com
--
-- Functions provide functionality of a musical instrument synthesizer or for Ukrainian speech synthesis
-- especially for poets, translators and writers.
--

module Melodics.Ukrainian (
  appendS16LEFile
  , convertToProperUkrainian
  , takeData
  , isUkrainian
  -- ** Since 0.9.0.0 version -- transformation function.
  , new2OldRepresentation
) where

import qualified Data.Vector.Unboxed as V
import qualified Data.Vector as VB
import qualified Data.ByteString.Char8 as B
import System.IO
import CaseBi.Unboxed (getBFst')
import qualified CaseBi as X (getBFst')
import qualified Melodics.ByteString.Ukrainian as MU (convertToProperUkrainianS)
import Paths_mmsyn6ukr

{-
-- Inspired by: https://mail.haskell.org/pipermail/beginners/2011-October/008649.html
-}

-- | Function to take raw sound data from the \".wav\" file given.
takeData :: FilePath -> IO B.ByteString
takeData file = do
  data1 <- B.readFile file
  let dataN = B.drop 44 data1 in return dataN

-- | The function that actually produces a .raw file. The mapping table is given in the @Map.txt@ file.
appendS16LEFile ::  VB.Vector String -> Handle -> IO ()
appendS16LEFile xs hdl | not (VB.null xs) =
  do
    dataFileList <- mapM getDataFileName
      ["-.wav", "0.wav", "1.wav", "A.wav", "B.wav", "C.wav", "D.wav", "E.wav", "F.wav", "G.wav", "H.wav",
        "I.wav", "J.wav", "K.wav", "L.wav", "M.wav", "N.wav", "O.wav", "P.wav", "Q.wav", "R.wav",
          "S.wav", "T.wav", "U.wav", "V.wav", "W.wav", "X.wav", "Y.wav", "Z.wav", "a.wav", "b.wav", "c.wav",
            "d.wav", "e.wav", "f.wav"]
    dataList <- VB.mapM takeData . VB.fromList $! dataFileList
    VB.mapM_ (\u ->
      if VB.all (\z -> B.length z > 0) dataList
        then let rs =  tail . dropWhile (/= ' ') . takeWhile (/= '}') . show $ hdl in do
          hClose hdl
          closedHdl <- hIsClosed hdl
          if closedHdl
            then do
                   B.appendFile rs $ dataList VB.! (X.getBFst' (0, VB.fromList [("-", 0), ("0", 1), ("1", 2), ("а", 3), ("б", 4),
                     ("в", 5), ("г", 6), ("д", 7), ("дж", 8), ("дз", 9), ("е", 10), ("ж", 11), ("з", 12), ("и", 13),
                        ("й", 14), ("к", 15), ("л", 16), ("м", 17), ("н", 18), ("о", 19), ("п", 20), ("р", 21),
                          ("с", 22), ("сь", 23), ("т", 24), ("у", 25), ("ф", 26), ("х", 27), ("ц", 28), ("ць", 29), ("ч", 30),
                            ("ш", 31), ("ь", 32), ("і", 33), ("ґ", 34)]) u)
            else error "File is not closed!"
        else error "Data sound file is not read!") xs
    hClose hdl
                       | otherwise = return ()

-- | The function that converts a written Ukrainian text into the sounding in the program phonetical respesentation.
-- It is not exact phonetically but you can make for yourself a general impression of the Ukrainian sounding.
convertToProperUkrainian :: String -> VB.Vector String
convertToProperUkrainian = new2OldRepresentation . MU.convertToProperUkrainianS

new2OldRepresentation :: String -> VB.Vector String
new2OldRepresentation xs = VB.fromList . map f $ xs
  where f = X.getBFst' ("",VB.fromList . zip "-01ABCDEFabcdefghijklmnopqrstuvwxyz" $ ["-","0","1","дз","ж","й","сь","ч","ш","а","б",
              "ц","д","е","ф","ґ","г","і","дж","к","л","м","н","о","п","ь","р","с","т","у","в","ць","х","и","з"])

isUkrainian :: Char -> Bool
isUkrainian y | (y >= '\1040' && y <= '\1065') || (y >= '\1070' && y <= '\1097') = True
              | otherwise = getBFst' (False, V.fromList . map (\x -> (x, True)) $ "'-\700\1028\1030\1031\1068\1100\1102\1103\1108\1110\1111\1168\1169\8217") y
